Refactor Papercraft API

noteflakes opened the pull request:
This PR changes the Papercraft API to remove the `Proc#render`, `Proc#apply` and similar methods, and instead add `Papercraft#render`, `Papercraft#apply` etc.

## Discussion

Some objections have been made to the fact that Papercraft extends the `Proc` class with a bunch of methods, and most importantly `#render` and `#apply`. This PR introduces a different API, where those methods are removed, and instead to render templates we need to use `Papercraft.render`, `Papercraft.apply` etc:

```ruby
templ = ->(name) { h1 name }

# before:
templ.render('foo')

# after:
Papercraft.render(templ, 'foo')

# same for apply
layout = -> (**props) {
  html {
    body {
      render_children(**props)
    }
  }
}

page = Papercraft.apply(layout) { |title:, **props| h1 title }
Papercraft.render(page, title: 'foo')
```

Note that Papercraft still extends Proc with some methods needed for storing and accessing template-specific metadata related to compilation and cached rendering. Those methods are prefixed and postfixed with double underscores. Those methods are used internally, and are not called directly by application code.

## Release

If merged, this will be released as version 3.0.

noteflakes committed:
Move render_xxx, apply methods to Papercraft module

Remove the following from the ProcExtensions module and make them singleton methods on Papercraft module:
- `render`
- `render_xml`
- `apply`
- `render_cache`

timriley commented on lib/papercraft/proc_ext.rb:
```diff
@@ -30,54 +30,11 @@ def __compiled_proc__(mode: :html)
-      @render_cache ||= {}
-      @render_cache[key] ||= render(*args, **kargs, &block)
+    # @return [Hash] cache hash
+    def __render_cache__
```
Might it be worth naming these e.g. `__papercraft_render_cache__`? That way you guarantee there there should never be a naming collision, and people introspecting innocent proc objects will know what these methods are for.

noteflakes replied:
Done: [d7eafb0](https://github.com/digital-fabric/papercraft/pull/21/commits/d7eafb06960357e13feaefce436778cdf4c17c65)

noteflakes committed:
Prefix internal Proc methods with __papercraft

noteflakes committed:
Improve Papercraft::Template wrapper class

- Accept block in `#initialize`.
- Add `#call` as alias to `#render`

noteflakes committed:
Add support for Papercraft.render { ... } form

Add more tests for Papercraft singleton methods

noteflakes committed:
Streamline Papercraft API

Use just the methods `html` and `xml`, and corresponding `cache_html`, `cache_xml` methods.

noteflakes committed:
Add support for self-closing XML tags

noteflakes committed:
Improve implementation of Papercraft.apply, more tests

noteflakes committed:
Version 3.0
